pkg/storagepoolclaim/v1alpha1/storagepoolclaim.go

```diff
@@ -0,0 +1,240 @@
+package v1alpha1
+
+import (
+	"text/template"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type annotationKey string
+
+const scheduleOnHostAnnotation annotationKey = "volume.kubernetes.io/selected-node"
+
+type spc struct {
+	// actual spc object
+	object *apis.StoragePoolClaim
+}
+
+// SPCList holds the list of storagepoolclaims
+type SPCList struct {
+	// list of storagepoolclaims
+	items []*spc
+}
+
+// predicate defines an abstraction
+// to determine conditional checks
+// against the provided spc instance
+type predicate func(*spc) bool
+
+type predicateList []predicate
+
+// all returns true if all the predicates
+// succeed against the provided csp
+// instance
+func (l predicateList) all(c *spc) bool {
+	for _, pred := range l {
+		if !pred(c) {
+			return false
+		}
+	}
+	return true
+}
+
+// IsNotUID returns true if provided csp
+// instance's UID does not match with any
+// of the provided UIDs
+func IsNotUID(uids ...string) predicate {
+	return func(c *spc) bool {
+		for _, uid := range uids {
+			if uid == string(c.object.GetUID()) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+// HasAnnotation returns true if provided annotation
+// key and value are present in the provided CSP
+// instance
+func HasAnnotation(key, value string) predicate {
+	return func(c *spc) bool {
+		val, ok := c.object.GetAnnotations()[key]
+		if ok {
+			return val == value
+		}
+		return false
+	}
+}
+
+// Filter will filter the csp instances
+// if all the predicates succeed against that
+// csp.
+func (l *SPCList) Filter(p ...predicate) *SPCList {
+	var plist predicateList
+	plist = append(plist, p...)
+	if len(plist) == 0 {
+		return l
+	}
+
+	filtered := ListBuilder().List()
+	for _, spc := range l.items {
+		if plist.all(spc) {
+			filtered.items = append(filtered.items, spc)
+		}
+	}
+	return filtered
+}
+
+type Builder struct {
+	object *apis.StoragePoolClaim
+}
+
+func NewSPCBuilder() *Builder {
+	return &Builder{
+		object: &apis.StoragePoolClaim{},
+	}
+}
+
+func (sb *Builder) WithSPC(spc *apis.StoragePoolClaim) *Builder {
+	sb.object = spc
+	return sb
+}
+
+func (sb *Builder) WithName(name string) *Builder {
```

Comment:
IMO, It is good to check length of name and return error if it is empty. (this kind of checks needed wherever it required)